catris.py

```diff
@@ -324,33 +363,442 @@ def move_blocks_down(self) -> set[Player]:
+                b"o--", b"o" + (COLOR % colors["s"]) + b"--"
+            )
+            template_line = template_line.replace(b"--o", b"--" + (COLOR % 0) + b"o")
+            if template_line.startswith(b"|"):
+                template_line = (
+                    (COLOR % colors["a"]) + b"|" + (COLOR % 0) + template_line[1:]
+                )
+            if template_line.endswith(b"|"):
+                template_line = (
+                    template_line[:-1] + (COLOR % colors["d"]) + b"|" + (COLOR % 0)
+                )
+
+            result_line = b""
+            while template_line:
+                if template_line[0] in b"wasd":
+                    letter = template_line[:1].decode("ascii")
+                    result_line += (
+                        (COLOR % colors[letter])
+                        + wrapped_names[letter].pop(0).encode("utf-8")
+                        + (COLOR % 0)
+                    )
+                    template_line = template_line.replace(template_line[:1], b"")
+                else:
+                    result_line += template_line[:1]
+                    template_line = template_line[1:]
+            result.append(result_line)
+
+        return result
+
+    def reset(self) -> None:
+        super().reset()
+        self.landed_blocks = {
+            (x, y): None
+            for x in range(-self.GAME_RADIUS, self.GAME_RADIUS + 1)
+            for y in range(-self.GAME_RADIUS, self.GAME_RADIUS + 1)
+            if max(abs(x), abs(y)) > self.MIDDLE_AREA_RADIUS
+        }
+
+    def wipe_playing_area(self, player: Player) -> None:
+        for x, y in self.landed_blocks.keys():
+            # Math magic to check if (x,y) is in the player's triangle-shaped area.
+            # Have fun figuring out how it works :)
+            dot = x * player.direction_x + y * player.direction_y
+            if dot >= 0 and 2 * dot ** 2 >= x * x + y * y:
+                self.landed_blocks[x, y] = None
+        player.moving_block_or_wait_counter = MovingBlock(player)
+
+    def is_valid(self) -> bool:
+        assert self.landed_blocks.keys() == {
+            (x, y)
+            for x in range(-self.GAME_RADIUS, self.GAME_RADIUS + 1)
+            for y in range(-self.GAME_RADIUS, self.GAME_RADIUS + 1)
+            if max(abs(x), abs(y)) > self.MIDDLE_AREA_RADIUS
+        }
+
+        if not super().is_valid():
+            return False
+
+        for block in self._get_moving_blocks():
+            for x, y in block.get_coords():
+                if max(abs(x), abs(y)) <= self.MIDDLE_AREA_RADIUS:
+                    return False
+                player_x, player_y = block.player.world_to_player(x, y)
+                if (
+                    player_x < -self.GAME_RADIUS
+                    or player_x > self.GAME_RADIUS
+                    or player_y > 0
+                ):
+                    return False
+        return True
+
+    def find_full_lines(self) -> list[int]:
```

Comment:
These are actualy rings, not lines. Comment exists about this but should copy here